src/v/kafka/server/quota_manager.cc

```diff
@@ -204,46 +202,62 @@ void quota_manager::update_client_quotas() {
 
 // If client is part of some group then client quota ID is a group
 // else client quota ID is client_id
-static std::optional<std::string_view> get_client_quota_id(
+static tracker_key get_client_quota_id(
   const std::optional<std::string_view>& client_id,
   const std::unordered_map<ss::sstring, config::client_group_quota>&
     group_quota) {
     if (!client_id) {
-        return std::nullopt;
+        // requests without a client id are grouped into an anonymous group that
+        // shares a default quota. the anonymous group is keyed on empty string.
+        return k_client_id{""};
     }
     for (const auto& group_and_limit : group_quota) {
         if (client_id->starts_with(
               std::string_view(group_and_limit.second.clients_prefix))) {
-            return group_and_limit.first;
+            return k_group_name{group_and_limit.first};
         }
     }
-    return client_id;
+    return k_client_id{*client_id};
 }
 
-int64_t quota_manager::get_client_target_produce_tp_rate(
-  const std::optional<std::string_view>& quota_id) {
-    if (!quota_id) {
-        return _default_target_produce_tp_rate();
-    }
-    auto group_tp_rate = _target_produce_tp_rate_per_client_group().find(
-      ss::sstring(quota_id.value()));
-    if (group_tp_rate != _target_produce_tp_rate_per_client_group().end()) {
-        return group_tp_rate->second.quota;
-    }
-    return _default_target_produce_tp_rate();
+int64_t
+quota_manager::get_client_target_produce_tp_rate(const tracker_key& quota_id) {
+    return std::visit(
+      [this](const auto& qid) -> int64_t {
+          using T = std::decay_t<decltype(qid)>;
+          if constexpr (std::is_same_v<k_client_id, T>) {
+              return _default_target_produce_tp_rate();
+          } else if constexpr (std::is_same_v<k_group_name, T>) {
+              auto group = _target_produce_tp_rate_per_client_group().find(qid);
+              if (group != _target_produce_tp_rate_per_client_group().end()) {
+                  return group->second.quota;
+              }
+              return _default_target_produce_tp_rate();
+          } else {
+              static_assert(always_false_v<T>, "Unknown tracker_key type");
+          }
+      },
+      quota_id);
```

Comment:
`ss::visit` is quite nice for this scenario:
```suggestion
    return ss::visit(
      quota_id,
      [this](const k_client_id&) -> int64_t {
          return _default_target_produce_tp_rate();
      },
      [this](const k_group_name& qid) -> int64_t {
          auto group = _target_produce_tp_rate_per_client_group().find(qid);
          if (group != _target_produce_tp_rate_per_client_group().end()) {
              return group->second.quota;
          }
          return _default_target_produce_tp_rate();
      });

Reply:
Nice, done